xarray/core/dataset.py

```diff
@@ -1433,11 +1449,68 @@ def __setitem__(self, key: Hashable, value) -> None:
         variable.
         """
         if utils.is_dict_like(key):
-            raise NotImplementedError(
-                "cannot yet use a dictionary as a key to set Dataset values"
-            )
+            from .dataarray import DataArray
 
-        self.update({key: value})
+            # check for consistency first
```

Comment:
These error checks are great!

There are enough of them that I would suggest moving them all into a helper method, so that the overall structure of `__setitem__`  is clearer.

Reply:
ok